torch/lib/c10d/ProcessGroupNCCL.cpp

```diff
@@ -148,37 +148,67 @@ const std::exception& ProcessGroupNCCL::WorkNCCL::exception() const {
       "isCompleted() and wait() will either succeed or throw");
 }
 
+std::unordered_map<ssize_t, ssize_t> ProcessGroupNCCL::pgUniqueNCCLIDCnt_;
+ssize_t ProcessGroupNCCL::processGroupCounter_ = -1;
+std::mutex ProcessGroupNCCL::pgTrackingLock_;
+
 ProcessGroupNCCL::ProcessGroupNCCL(
     const std::shared_ptr<Store>& store,
     int rank,
     int size)
     : ProcessGroup(rank, size), store_(store) {
-  C10D_CUDA_CHECK(cudaGetDeviceCount(&numGPUs_));
   thcState_ = ::at::globalContext().lazyInitCUDA();
+  // Generate the Process Group ID for current PG, this needs to be identical
+  // for all processes
+  std::unique_lock<std::mutex> lock(pgTrackingLock_);
+  while (true) {
+    ++processGroupCounter_;
+    if (pgUniqueNCCLIDCnt_.find(processGroupCounter_) ==
+        pgUniqueNCCLIDCnt_.end()) {
+      pgUniqueNCCLIDCnt_[processGroupCounter_] = -1;
+      processGroupID_ = std::to_string(processGroupCounter_);
+      return;
+    }
+  }
```

Comment:
I made a comment about this in an earlier revision; if you already hold the lock, why do you need the loop? If you just increment the process group counter that should never conflict with anything else. Even better, it could be an atomic int and you don't even need the lock. Then with that value you set the map entry to -1 (where you do need the lock of course) and are done, without needing the while loop.

Reply:
I don't simply want to increase the counter and overflow, that's what the while loop is for. And for the lock is self, it is needed since we increase the counter and then read it, and then use it to  access the map, the whole op needs to be atomic.

Reply:
Why would it overflow?